### PR TITLE
Android demo app for neural machine translation: code, README, screenshots

### PR DESCRIPTION
**Pull request by jeffxtang**

### CODE REVIEW

File: Seq2SeqNMT/README.md [hunk 1]
```diff
@@ -0,0 +1,38 @@
...
+
+* The Python code that saves the encoder and attention-enhanced decoder models, trained based on the code in the PyTorch NMT tutorial;
+* The Python code that loads the saved encoder and decoder models, optimizes them and saves again for mobile apps to use;
+* The Android app that uses the encoder and decoder models to do French-English translation.
+
+## Prerequisites
+
+* PyTorch 1.7 or later (Optional)
+* Python 3.8 (Optional)
+* Android Pytorch library 1.6, 1.7 or later
```
Comment (IvanKobzarev) [inline Seq2SeqNMT/README.md:17]: If you provide links to the pt models that can be download - it is better to specify one version that was used to script them. (As serialization format may change between versions)

File: Seq2SeqNMT/app/build.gradle [hunk 2]
```diff
@@ -0,0 +1,39 @@
...
+        versionCode 1
+        versionName "1.0"
+
+        testInstrumentationRunner "androidx.test.runner.AndroidJUnitRunner"
+    }
+
+    buildTypes {
+        release {
+            minifyEnabled false
+            proguardFiles getDefaultProguardFile('proguard-android-optimize.txt'), 'proguard-rules.pro'
```
Comment (IvanKobzarev) [inline Seq2SeqNMT/app/build.gradle:20]: default 'proguard-rules.pro' has all the lines commented out - maybe to delete it and this line?

File: Seq2SeqNMT/app/src/main/java/org/pytorch/demo/seq2seqnmt/MainActivity.java [hunk 3]
```diff
@@ -0,0 +1,184 @@
...
+import java.nio.FloatBuffer;
+import java.nio.LongBuffer;
+import java.util.ArrayList;
+
+public class MainActivity extends AppCompatActivity implements Runnable{
+    // to be consistent with the model inputs defined in seq2seq_nmt.py, based on
+    // https://pytorch.org/tutorials/intermediate/seq2seq_translation_tutorial.html
+    private final int HIDDEN_SIZE = 256;
+    private final int EOS_token = 1;
+    private final int MAX_LENGTH = 50;
```
Comment (IvanKobzarev) [inline Seq2SeqNMT/app/src/main/java/org/pytorch/demo/seq2seqnmt/MainActivity.java:35]: make static and capitalize?
    private static final int HIDDEN_SIZE = 256;
    private static final int EOS_TOKEN = 1;
    private static final int MAX_LENGTH = 50;

File: Seq2SeqNMT/app/src/main/java/org/pytorch/demo/seq2seqnmt/MainActivity.java [hunk 4]
```diff
@@ -0,0 +1,184 @@
...
+
+        mButton = findViewById(R.id.btnTranslate);
+        mEditText = findViewById(R.id.etFrom);
+        mTextView = findViewById(R.id.tvTo);
+
+        mButton.setOnClickListener(new View.OnClickListener() {
+            public void onClick(View v) {
+            mButton.setEnabled(false);
+            Thread thread = new Thread(MainActivity.this);
+            thread.start();
```
Comment (IvanKobzarev) [inline Seq2SeqNMT/app/src/main/java/org/pytorch/demo/seq2seqnmt/MainActivity.java:59]: Indent

File: Seq2SeqNMT/app/src/main/java/org/pytorch/demo/seq2seqnmt/MainActivity.java [hunk 5]
```diff
@@ -0,0 +1,184 @@
...
+            is = getAssets().open("source_wrd2idx.json");
+            size = is.available();
+            buffer = new byte[size];
+            is.read(buffer);
+            is.close();
+
+            json = new String(buffer, "UTF-8");
+            wrd2idx = new JSONObject(json);
+        } catch (JSONException | IOException e) {
+            e.printStackTrace();
```
Comment (IvanKobzarev) [inline Seq2SeqNMT/app/src/main/java/org/pytorch/demo/seq2seqnmt/MainActivity.java:103]: By default printStackTrace prints to System.err
it is not shown in logcat by default,
I think it is better to have it in logcat, smth like:

    android.util.Log.e(TAG, "Error reading json asset", e)

File: Seq2SeqNMT/app/src/main/java/org/pytorch/demo/seq2seqnmt/MainActivity.java [hunk 6]
```diff
@@ -0,0 +1,184 @@
...
+            mInputTensor = Tensor.fromBlob(mInputTensorBuffer, decoderInputShape);
+        }
+
+        String english = "";
+        try {
+            for (int i = 0; i < result.size(); i++)
+                english += " " + idx2wrd.getString("" + result.get(i));
+        }
+        catch (JSONException e) {
+            e.printStackTrace();
```
Comment (IvanKobzarev) [inline Seq2SeqNMT/app/src/main/java/org/pytorch/demo/seq2seqnmt/MainActivity.java:180]: same
android.util.Log.e(TAG, "", e)

File: Seq2SeqNMT/app/src/main/java/org/pytorch/demo/seq2seqnmt/MainActivity.java [hunk 7]
```diff
@@ -0,0 +1,184 @@
...
+
+        String french = text;
+        long[] inputs =  new long [french.split(" ").length];
+        try {
+            for (int i = 0; i < french.split(" ").length; i++) {
+                inputs[i] = wrd2idx.getLong(french.split(" ")[i]);
+            }
+        }
+        catch (JSONException e) {
+            e.printStackTrace();
```
Comment (IvanKobzarev) [inline Seq2SeqNMT/app/src/main/java/org/pytorch/demo/seq2seqnmt/MainActivity.java:115]: same
android.util.Log.e(TAG, "", e)

**Review by IvanKobzarev — APPROVED** [review]

File: Seq2SeqNMT/README.md (hunk 1 above)
Comment (jeffxtang) [reply Seq2SeqNMT/README.md:17]: Great point! Will do.

**Pull request merged**